preprocess.py

```diff
@@ -0,0 +1,176 @@
+from FixNamesTransformer import FixNamesTransformer
+# from DropSmallMissingData import DropSmallMissingData
+from sklearn.compose import ColumnTransformer
+from sklearn.pipeline import Pipeline
+import pandas as pd
+import numpy as np
+import re
+from sklearn.preprocessing import OrdinalEncoder, OneHotEncoder, FunctionTransformer
+from sklearn.linear_model import LinearRegression
+from sklearn.experimental import enable_iterative_imputer  # IterativeImputer doesn't work without this import
+from sklearn.impute import IterativeImputer
+from feature_engine.selection import DropFeatures
+from typing import List, Tuple
+import warnings
+warnings.filterwarnings("ignore")
+
+
+def drop_small_missing_data(df: pd.DataFrame, features: List[str]) -> pd.DataFrame:
+    nulls_percent = df.isna().sum() / len(df)
+    # dropping row for features with less than 0.5% of missing values
+    small_missing_values_features = [feature for feature in features if 0 < nulls_percent[feature] < 0.005]
+    for feature in small_missing_values_features:
+        df = df[df[feature].notnull()]
+    return df
+
+
+def feature_index_for_missing_data(df: pd.DataFrame, features: List[str]) -> pd.DataFrame:
+    nulls_count = df.isna().sum()
+    for feature in features:
+        if nulls_count[feature] >= 1:
+            # a binary column indicating which rows have missing data for that feature
+            if feature == 'Survived':  # wouldn't happen (full data on independent variable)
+                continue
+            missing_feature_name = f"{feature}NaN"
+            df[missing_feature_name] = np.where(df[feature].isnull(), 1, 0)
+    return df
+
+
+def find_honorifics(name: str) -> str:
+    honorific_names = r'(?:Mrs|Mr|Ms|Miss|Master|Don|Rev|Mme|Major|Mlle|Col|Capt|Jonkheer|Countess|Dr)\.?'
+    honorific = re.findall(honorific_names, name)
+    if not honorific:  # empty list - no honorific found
+        return 'non_honorific'
+    honorific = honorific[0]  # findall returns a list - in our case list of single string
+    honorific = honorific.split()[0]  # the string contains the honorific in first place if exists
+    honorific = honorific.replace('.', '')
+    return honorific
+
+
+def honorifics_feature_from_name(df: pd.DataFrame) -> pd.DataFrame:
+    df['Honorifics'] = df['Name'].apply(lambda name: find_honorifics(name))
+    return df
+
+
+def solo_feature_from_parch_sibsp(df, y=None) -> pd.DataFrame:
+    # If both SibSp and Parch are 0 then we say the passenger is by himself on board, meaning he is traveling solo
+    df['Solo'] = (df['Parch'] + df['SibSp']).apply(lambda x: 1 if x == 0 else 0)
+    return df
+
+
+def CabinChar_feature_from_cabin(df: pd.DataFrame) -> pd.DataFrame:
```

Comment:
functions name should NEVER start with capital letter!!!!!!!
you can use camel case or snake case